File: backend/features/flashcards/parsing.py
"""Turning text into cards, by the two routes that do not involve a model.

``parse_qa_pairs`` reads the ``Q::``/``A::`` convention the vault already uses
— every note ingest writes carries a self-test tail in exactly this shape
(``backend/agent/prompts/note_quality.md``), which is what makes "import from
a note" work against any note rather than only against ``flashcards.md``.

``parse_delimited`` reads pasted rows, the affordance every flashcard tool
offers: choose what separates the two halves and what separates the rows,
paste, preview. ``web/lib/flashcards/parsing.ts`` is its twin, so the preview
shown before importing is the same parse the server will perform.
"""

from __future__ import annotations

#: What may separate a card's two halves. Keys are the wire values.
FIELD_DELIMITERS: dict[str, str] = {"tab": "\t", "comma": ",", "dash": "-"}

#: What may separate one card from the next.
ROW_DELIMITERS: dict[str, str] = {"newline": "\n", "semicolon": ";"}


def parse_qa_pairs(text: str) -> list[tuple[str, str]]:
    """Parse ``Q:: <front>`` / ``A:: <back>`` pairs from markdown.

    Each field may continue on following lines up to the next ``Q::`` marker
    (or end of text). Pairs missing either half are dropped.
    """
    pairs: list[tuple[str, str]] = []
    front_lines: list[str] | None = None
    back_lines: list[str] | None = None
    mode: str | None = None

    def flush() -> None:
        if front_lines is not None and back_lines is not None:
            front = "\n".join(front_lines).strip()
            back = "\n".join(back_lines).strip()
            if front and back:
                pairs.append((front, back))

    for line in text.splitlines():
        if line.startswith("Q::"):
            flush()
            front_lines = [line[3:].strip()]
            back_lines = None
            mode = "q"
        elif line.startswith("A::"):
            back_lines = [line[3:].strip()]
            mode = "a"
        elif mode == "q" and front_lines is not None:
            front_lines.append(line)
        elif mode == "a" and back_lines is not None:
            back_lines.append(line)
    flush()
    return pairs


def parse_delimited(text: str, *, field: str, row: str) -> list[tuple[str, str]]:
    """Parse pasted rows into ``(front, back)`` pairs.

    Splits on the **first** field delimiter only. A definition legitimately
    contains commas and hyphens, and splitting greedily would truncate exactly
    the cards worth writing.

    Rows lacking a delimiter, or with either half empty, are dropped rather
    than imported half-formed — a card with no answer is not a card, and the
    preview counts what will actually be created.

    Raises ``KeyError`` for an unknown delimiter name; these arrive from a
    request body and are not trusted input.
    """
    field_sep = FIELD_DELIMITERS[field]
    row_sep = ROW_DELIMITERS[row]
    pairs: list[tuple[str, str]] = []
    for line in text.replace("\r\n", "\n").split(row_sep):
        head, sep, tail = line.partition(field_sep)
        if not sep:
            continue
        front, back = head.strip(), tail.strip()
        if front and back:
            pairs.append((front, back))
    return pairs

File: web/lib/flashcards/parsing.ts
/**
 * The browser twin of `backend/features/flashcards/parsing.py::parse_delimited`.
 *
 * It exists so the import dialog can show what *will* be created before
 * anything is sent. That only helps if the two agree exactly, so the rules
 * here are the same rules, and the unit tests mirror `test_parsing.py`
 * case for case.
 */

export const FIELD_DELIMITERS: Record<string, string> = {
  tab: "\t",
  comma: ",",
  dash: "-",
};

export const ROW_DELIMITERS: Record<string, string> = {
  newline: "\n",
  semicolon: ";",
};

export interface ParsedCard {
  front: string;
  back: string;
}

export type ImportFormat = "qa" | "delimited";

export interface Detected {
  format: ImportFormat;
  field: string;
  row: string;
  cards: ParsedCard[];
}

/**
 * Parse `Q:: … / A:: …` pairs, the shape every note Argus generates carries in
 * its self-test section.
 *
 * A browser twin of `backend/features/flashcards/parsing.py::parse_qa_pairs`,
 * kept only so a dropped file can be *previewed* before it is sent; the server
 * re-parses with the Python original, which stays authoritative. Continuation
 * lines are folded the same way: a field runs until the next `Q::` marker.
 */
export function parseQaPairs(text: string): ParsedCard[] {
  const cards: ParsedCard[] = [];
  let front: string[] | null = null;
  let back: string[] | null = null;
  let mode: "q" | "a" | null = null;

  const flush = () => {
    if (front === null || back === null) return;
    const f = front.join("\n").trim();
    const b = back.join("\n").trim();
    if (f && b) cards.push({ front: f, back: b });
  };

  for (const line of text.replace(/\r\n/g, "\n").split("\n")) {
    if (line.startsWith("Q::")) {
      flush();
      front = [line.slice(3).trim()];
      back = null;
      mode = "q";
    } else if (line.startsWith("A::")) {
      back = [line.slice(3).trim()];
      mode = "a";
    } else if (mode === "q" && front !== null) {
      front.push(line);
    } else if (mode === "a" && back !== null) {
      back.push(line);
    }
  }
  flush();
  return cards;
}

/**
 * Guess how a pasted or dropped body is laid out.
 *
 * `Q::` wins outright wherever it appears — a note carrying the self-test tail
 * is unambiguous, and a file that has both markers *and* tabs is far more
 * likely to be prose with an indented line than a two-column table.
 *
 * Otherwise every field/row delimiter pairing is tried and the one yielding
 * the most cards wins, with ties broken by the order in `FIELD_DELIMITERS` so
 * the result is deterministic rather than dependent on object iteration.
 *
 * The guess is only ever a starting point: the caller shows it and lets it be
 * overridden. A detector that cannot be corrected is worse than no detector,
 * because a wrong guess is then indistinguishable from a broken file.
 */
export function detectFormat(text: string): Detected {
  const qa = parseQaPairs(text);
  if (qa.length > 0) return { format: "qa", field: "tab", row: "newline", cards: qa };

  let best: Detected = { format: "delimited", field: "tab", row: "newline", cards: [] };
  for (const field of Object.keys(FIELD_DELIMITERS)) {
    for (const row of Object.keys(ROW_DELIMITERS)) {
      const cards = parseDelimited(text, field, row);
      if (cards.length > best.cards.length) {
        best = { format: "delimited", field, row, cards };
      }
    }
  }
  return best;
}

/**
 * Parse pasted rows into cards.
 *
 * Splits on the **first** field delimiter only: a definition legitimately
 * contains commas and hyphens, and splitting greedily would truncate exactly
 * the cards worth writing. Rows with no delimiter, or with either half empty,
 * are dropped rather than imported half-formed — so the count this returns is
 * the count the user will get.
 *
 * An unknown delimiter name yields no cards rather than throwing; the dialog
 * only ever passes names it got from the server.
 */
export function parseDelimited(text: string, field: string, row: string): ParsedCard[] {
  const fieldSep = FIELD_DELIMITERS[field];
  const rowSep = ROW_DELIMITERS[row];
  if (fieldSep === undefined || rowSep === undefined) return [];

  const cards: ParsedCard[] = [];
  for (const line of text.replace(/\r\n/g, "\n").split(rowSep)) {
    const at = line.indexOf(fieldSep);
    if (at === -1) continue;
    const front = line.slice(0, at).trim();
    const back = line.slice(at + fieldSep.length).trim();
    if (front && back) cards.push({ front, back });
  }
  return cards;
}
